### PR TITLE
GAP-2140: UI for view & manage editors

### PR DESCRIPTION
**Pull request by john-tco**
## Description

Ticket # and link

https://technologyprogramme.atlassian.net/jira/software/projects/GAP/boards/216?selectedIssue=GAP-2140

Adds ui for viewing and managing editors

Screenshot 2024-02-26 at 10 45 20

Screenshot 2024-02-26 at 10 44 56

Screenshot 2024-02-26 at 10 44 21

Screenshot 2024-02-26 at 10 46 59

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [x] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [x] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

### CODE REVIEW

File: packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+  const editorRows = (
+    await getSchemeEditors(schemeId, sessionCookie, userServiceJwt)
+  ).map(formatEditorRows);
+
+  editorRows.unshift({
+    key: 'Email',
+    value: 'Role',
+    action: isOwner ? 'Actions' : '',
+  } as UnformattedEditorRow);
```
[the remark this answers is not in the document]
Comment (dominicwest) [reply packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx:69]: There is also _a lot_ of logic thats formatting the data for this table. We have the following functions:
- formatEditorRows
- formatEditorRow

and logic in between all that too:

```
  const editorRows = schemeEditors.map(formatEditorRows);

  editorRows.unshift({
    key: 'Email',
    value: 'Role',
    action: isOwner ? 'Actions' : '',
  });

  const mappedTableRows = editorRows.reduce(formatEditorRow, []);
```
I'm finding it very hard to follow what its actually doing and I suspect this can be cleaned up a lot

File: packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  ariaLabel: string;
+};
+
+export type UnformattedEditorRow = {
+  key: string;
+  value: string;
+  action?: string | Action | JSX.Element;
+};
+
+const getEditorsServerSideProps = async ({
```
Comment (dominicwest) [inline packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx:30]: can also just call this getServerSideProps and export it directly rather then `export const getServerSideProps = async (ctx: GetServerSidePropsContext) =>
  runGetEditorsAndHandleError(ctx);`

File: packages/admin/src/services/SchemeService.ts [hunk 3]
```diff
@@ -62,6 +64,26 @@ export const getGrantScheme = async (schemeId: string, sessionId: string) => {
...
+  schemeId: string,
+  sessionId: string,
+  userServiceJwt: string
+) => {
+  const { data } = await axios.get<EditorList[]>(
+    `${SCHEME_EDITORS_HOST}/${schemeId}`,
+    getFullConfig(sessionId, userServiceJwt)
+  );
+  return data;
+};
```
Comment (dominicwest) [inline packages/admin/src/services/SchemeService.ts:85]: Shift to `editorService`

File: packages/admin/src/utils/session.ts [hunk 4]
```diff
@@ -35,6 +35,13 @@ export const axiosSessionConfig = (sessionId: string) => {
   };
 };
 
+export const getFullConfig = (sessionId: string, userServiceJwt: string) => ({
```
Comment (dominicwest) [inline packages/admin/src/utils/session.ts:38]: Nice I rate this <3

File: packages/gap-web-ui/src/components/summary-list/SummaryList.tsx [hunk 5]
```diff
@@ -15,59 +17,77 @@ export interface Row {
   action?: string | JSX.Element;
 }
 
+const STYLE_MAP = {
+  displayRegularKeyFont: 'gap-summary-list--key-weight-regular',
+  hasWiderKeyColumn: 'key-width-45percent-sm',
+  equalColumns: 'equal-columns',
+  useBoldFont: 'fw-bold',
+  actionTextLeft: 'text-left',
+};
```
Comment (dominicwest) [inline packages/gap-web-ui/src/components/summary-list/SummaryList.tsx:26]: I miss tailwind

**Review by dominicwest — CHANGES_REQUESTED** [review]

File: packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (john-tco) [reply packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx:69]: The formatting logic here is just because the page provides data for both view and manage ui. In retrospect the api should probs handle this data formatting for us.

File: packages/gap-web-ui/src/components/summary-list/SummaryList.tsx (hunk 5 above)
Comment (john-tco) [reply packages/gap-web-ui/src/components/summary-list/SummaryList.tsx:26]: We would need to have this same pattern / style map with tailwind tbh, there's too many ui scenarios for this poor table component.

File: packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx (hunk 2 above)
Comment (john-tco) [reply packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx:30]: Don't see why this is better? There's a clear distinction between business logic and error handling code which we will lose if we do that.

File: packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  getGrantScheme,
+  getSchemeEditors,
+  isSchemeOwner,
+} from '../../../services/SchemeService';
+
+export type EditorList = {
+  role: 'EDITOR' | 'OWNER';
+  email: string;
+  id: string;
+};
```
[the remark this answers is not in the document]
Comment (john-tco) [reply packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx]: Sure, moved that

File: packages/admin/src/pages/scheme/[schemeId]/index.module.scss [hunk 7]
```diff
@@ -0,0 +1,10 @@
+.breakLine {
+  background-color: #1d70b8;
```
[the remark this answers is not in the document]
Comment (john-tco) [reply packages/admin/src/pages/scheme/[schemeId]/index.module.scss]: fixed

File: packages/admin/src/pages/scheme/[schemeId]/manage-editors/index.page.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  return [...acc, { ...row, action }];
+};
+
+const ManageEditors = ({
+  editorRows,
+  schemeName,
+  schemeId,
+}: InferProps<typeof getServerSideProps>) => {
+  const mappedTableRows = editorRows.reduce(formatEditorRow, []);
```
Comment (dominicwest) [inline packages/admin/src/pages/scheme/[schemeId]/manage-editors/index.page.tsx:42]: Correct me if I'm wrong but I think this evaluates to the same in all instances
```suggestion
const formatEditorRow = (row: UnformattedEditorRow) => {
  const mapRequired = typeof row.action !== 'string' && row.action;
  if (row.value === OWNER) return { ...row, action: '-' };
  if (mapRequired)
    return { ...row, action: <LinkToAction action={row.action as Action} /> };
  return row;
};

const ManageEditors = ({
  editorRows,
  schemeName,
  schemeId,
}: InferProps<typeof getServerSideProps>) => {
  const mappedTableRows = editorRows.map(formatEditorRow);
```

File: packages/admin/src/pages/scheme/[schemeId]/view-editors/index.page.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  ...acc,
+  { ...row, action: undefined },
+];
+
+const ViewEditors = ({
+  editorRows,
+  schemeName,
+  schemeId,
+}: InferProps<typeof getServerSideProps>) => {
+  const mappedTableRows = editorRows.reduce(rmAction, []);
```
Comment (dominicwest) [inline packages/admin/src/pages/scheme/[schemeId]/view-editors/index.page.tsx:23]:
```suggestion
const rmAction = (row: UnformattedEditorRow): Row => ({
  ...row,
  action: undefined,
});

const ViewEditors = ({
  editorRows,
  schemeName,
  schemeId,
}: InferProps<typeof getServerSideProps>) => {
  const mappedTableRows = editorRows.map(rmAction);
```

Comment (dominicwest) [inline packages/admin/src/pages/scheme/[schemeId]/view-editors/index.page.tsx:23]:
```suggestion

export { getServerSideProps };

const rmAction = (row: UnformattedEditorRow) => ({
  ...row,
  action: undefined,
});

const ViewEditors = ({
  editorRows,
  schemeName,
  schemeId,
}: InferProps<typeof getServerSideProps>) => {
  const mappedTableRows = editorRows.map(rmAction);
```

File: packages/admin/src/pages/scheme/[schemeId]/manage-editors/index.page.tsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
+import InferProps from '../../../../types/InferProps';
+import { SummaryList } from 'gap-web-ui';
+import Meta from '../../../../components/layout/Meta';
+import CustomLink from '../../../../components/custom-link/CustomLink';
+import {
+  Action,
+  UnformattedEditorRow,
+  getServerSideProps,
+} from '../editors.getServerSideProps';
+import { Row } from 'gap-web-ui/dist/cjs/components/summary-list/SummaryList';
```
Comment (dominicwest) [inline packages/admin/src/pages/scheme/[schemeId]/manage-editors/index.page.tsx:10]: I reckon this might break in sandbox - need to export this type from gap-web-ui and import properly

Comment (john-tco) [reply packages/admin/src/pages/scheme/[schemeId]/manage-editors/index.page.tsx:10]: oops, good spot

File: packages/admin/src/pages/scheme/[schemeId]/view-editors/index.page.tsx (hunk 9 above)
Comment (dominicwest) [reply packages/admin/src/pages/scheme/[schemeId]/view-editors/index.page.tsx:23]: Reduce is only needed over a map if you need to continuously use the previous map iterations response, here we only need the current element

**Review by dominicwest — APPROVED** [review]
